Reduce .text footprint of the network stack

Taiki-San opened the pull request:
### Description

Patches to LWIP and some code from netsocket was using `s[n]printf` and `sscanf` to generate and parse strings based on integers. Specifically, converting MAC addresses and IPv6 addresses to and from their hexadecimal form and IPv4 addresses to and from their printable decimal form.

The use of those functions in public APIs forced the compiler to bring the full parser necessary to make `*printf/sscanf` work. This parser is massive (in order to handle floating point for instance) and several of its components were in the largest functions in the final binary *independently*, despite the monolithics `tcp_input` and `tcp_output`. 

The small test program I was using had the following memory breakdown before linking to the network stack when compiling with GCC and a release profile:
```
Text 38.2KB Data 1.23KB BSS 9.21KB
ROM 39.5KB RAM 10.4KB
```
The same program using the default network stack (K64F & Ethernet):
```
Text 114KB Data 1.76KB BSS 54.4KB
ROM 116KB RAM 56.1KB
```
Then the same program using the network stack included in the PR, not using `*printf/sscanf` and thus not linking with the expensive parser:
```
Text 87.3KB Data 1.4KB BSS 54.3KB
ROM 88.7KB RAM 55.7KB
```

Although the RAM change is negligeable (but still down), the .text impact is massive (-23%).

|   | No network  | Default network | Patched network | `*printf` parser only |
| ------- | ----------- | --------------- | --------------- | ------------------- |
| ROM change | 1x | 3.03x | 2.24x | 0.71x |
| Overhead | 0kB | 76.5kB | 49.2kB | 27.3kB |

I couldn't find where to put the small utils I had to write (a basic itoa and two bin2hex) so they are inline and bin2hex had to be duplicated.

### Pull request type

- [ ] Fix
- [x] Refactor
- [ ] New target
- [ ] Feature
- [ ] Breaking change

adbridge commented:
@Taiki-San  This needs a rebase due to conflicts. Also could you please address the review comments. In order for this to make it into mbed-os-5.9, this needs to be reviewed/through ci and merged by this Friday.

Taiki-San commented:
Understood, I'll rebase shortly.  
As for the issues raised by @kjbracey-arm, not sure what the solution would be.
Although the core suggestion was to prevent the unecessary code to be linked when not actively used, further discussion demonstrated that it wouldn't be trivial to do/require API changes.
Moreover, considering the first point (that there is no point in avoiding `sprintf` if it'll be used by the application code) was more or less rendered moot (the helper functions introduced by this commit takes up less space than a call to sprintf, even if the application uses sprintf elsewhere), how should I address it?

cmonr commented:
@Taiki-San Please undo the last commit, and instead do a rebase of this branch.

@kjbracey-arm Now that the big emac push is complete, mind addressing @Taiki-San last comment/question?

Taiki-San commented:
Squashed and rebased the commits

cmonr commented:
@SeppoTakalo Looks like there was a target device that left the network? Mind letting us know if we can restart pr-head?

cmonr commented:
@Taiki-San Looks like this'll need another rebase. The file in question is triggering the build error on pr-head as well, so it's possible the rebase will make pr-head green.

SeppoTakalo reviewed (changes requested):
I'm not in favor of this.

It saves flash on simple applications that do not use `sprintf()` or `sscanf()`

If someone uses `sscanf()` then net result would be zero savings. Downside would be more code to maintain.

Taiki-San commented:
This was [demonstrated to be incorrect](https://github.com/ARMmbed/mbed-os/pull/6293#issuecomment-374568767). Even when the application is using `sprintf`, replacing this use in SocketAddress by a new function resulted in a net decrease of flash size. I'm expecting a similar behavior with `sscanf`.

They are also different functions and bring in significantly different code (only a subset of the floating point parsing code is shared). Removing all `sscanf` uses result in savings, removing all `sprintf` uses result in savings, removing both result in more savings. Moreover, although I agree uses of `sprintf` are common enough that it can't really be expected that applications won't use it, `sscanf` is substantially less common.

Finally, I understood that the goal of mbedOS was to run on as many IoT as possible, including highly constrained devices who may want to remove their uses of `sprintf` is order to save space, but won't get it back because of this code. Although this additionnal code will require maintenance (?), it lowers the required footprint of mbedOS significantly and is a possibly useful util to add to the standard library.

kjbracey commented:
Could I suggest another approach here? The conflict currently arising is that #7119 has already stripped out all the local IPv6 code to use the conversion code over in libservice/libip6string - that was primarily to improve the output formatting. That conversion code does not use `sprintf` or `sscanf`, but is likely to be a little bigger to get the better formatting.

Rather than mess with the local code here, could you extend libservice to offer direct IPv4 equivalents of those IPv6 functions? That would be useful for a separate pool of code based around libservice, both in and outside mbed OS.  The master repo for libservice is https://github.com/ARMmbed/nanostack-libservice

I believe avoiding `sscanf` is worthwhile,  more doubtful about `sprintf`, but given that `ip6tos` doesn't use it, an `ip4tos` that doesn't use it would at least be consistent...

If touching libservice, might be worth also adding the "is this an IPv4/IPv6 literal?" call that is still local here.

Taiki-San commented:
Sounds good, I'll have a look at it shortly.

kjbracey commented:
Super. I was wrong though - it does use `sprintf`, but not `sscanf`.

cmonr commented:
@Taiki-San Any progress on this PR?

Taiki-San commented:
Not yet, just came back from a conference. Planned on spending some time on it tomorrow.
Are you under some kind of time pressure?

cmonr commented:
@Taiki-San No worries. Part of our job is to make sure PRs don't get stale due to inactivity. If they go on without any updates for too long, we'll close them.

cmonr commented:
@Taiki-San Is this still being worked on?

This now needs a rebase as well.

Taiki-San commented:
We're merging changes to libservice so that the tools this PR is introducing will be stored somewhere else. I'll then replace the changes of this PR to use the newly introduced tools.
This is still moving, although the changes are happening in https://github.com/ARMmbed/nanostack-libservice/pull/69 for now.

cmonr commented:
@Taiki-San Is this PR now unblocked?

Taiki-San commented:
Waiting on @kjbracey-arm to review https://github.com/ARMmbed/nanostack-libservice/pull/72.
I could probably rebase already with a first batch of changes but I'd like to switch to using the new functions from both nanostack-libservice's PR in a single mbedOS PR.

cmonr commented:
Making a note here, @kjbracey-arm is still on vacation.

cmonr commented:
@kjbracey-arm Reminder that this is waiting for your feedback, when you get a chance.

@SeppoTakalo I take it you're still not in favor of this, even with the follow up comments?

@Taiki-San A rebase will be needed.

kjbracey commented:
Hi, afraid we won't be able to get ARMmbed/nanostack-libservice#72 in in time for mbed OS 5.10, due to the timing of the Nanostack/libservice release cycle, but you can create an initial pass for an mbed OS 5.10 NSAPI patch using the previous additions.

Taiki-San commented:
I think that by using libservice, this should mitigate @SeppoTakalo concerns. Not all the changes I was looking for were made but by using a more modular approach, the end goal is reached. Close enough for me.
@kjbracey-arm: Ok. So I may use the new APIs but not yet check `stoip6` for success/failure. Is that correct?

cmonr commented:
Is this PR still parked/waiting on https://github.com/ARMmbed/nanostack-libservice/pull/72 ?

Taiki-San committed:
Switch the IPv4 parser to the one recently introduced in libservice. Return a bool value on wether parsing of the IPv4 address string was successful

Taiki-San commented:
Sorry, this has been a busy month. I switched the IPv4 code (that was most of what was left) to the new function in libservice.
Moreover, because we are now able to report whether parsing of the string was successful, I changed the signature of `ipv4_from_address` to return said status via a bool. Is this change okay with everyone? If so, a new patch will be necessary to propagate this change to `ipv6_from_address` as the same change was made to the underlying parser in libservice.

kjbracey commented on features/netsocket/SocketAddress.cpp:
```diff
@@ -139,8 +116,7 @@ bool SocketAddress::set_ip_address(const char *addr)
         _addr.version = NSAPI_IPv4;
-        ipv4_from_address(_addr.bytes, addr);
-        return true;
+        return ipv4_from_address(_addr.bytes, addr);
```
Shouldn't this call `ipv4_from_address` straight away and check its result? You then don't need `ipv4_is_valid`.

kjbracey commented on features/netsocket/SocketAddress.cpp:
```diff
@@ -62,48 +63,24 @@ static bool ipv6_is_valid(const char *addr)
-                return;
-            }
-        }
+    if (addr == NULL)
```
Don't need the NULL check. No need to be defensive on an internal function, when we know we've already checked.

kjbracey reviewed (commented):
Looking neat. Fine with the return value addition, but use it more.

Taiki-San commented:
I forgot that `ipv4_from_address` was `static`, and feared external uses. Are those functions even still useful though? Considering they are merely thin wrappers around the libservice API?

kjbracey commented:
Probably not, no.

Taiki-San commented:
Alright, removed `ipv[4-6]_[to-from]_address`

0xc0170 commented:
> kjbracey-arm approved these changes 4 minutes ago

Is this now ready for review (one outstanding "requested changes") ? Please remove needs preceding PR if the nanostack PR was resolved (I can see it is still open).

kjbracey commented:
This uses the first preceding PR, which is already in.

It could be refined later when the second libservice PR goes in after 5.10. (Dropping `ipv6_is_valid`)

0xc0170 commented:
/morph build

Taiki-San commented:
What's the build failure cause?

0xc0170 commented:
> What's the build failure cause?

One of the servers went down, restarting

/morph build

cmonr commented:
/morph mbed2-build
